baseapp/state/manager.go

```diff
@@ -0,0 +1,152 @@
+package state
+
+import (
+	"fmt"
+	"sync"
+
+	cmtproto "github.com/cometbft/cometbft/proto/tendermint/types"
+
+	"cosmossdk.io/core/header"
+	"cosmossdk.io/log"
+	storetypes "cosmossdk.io/store/types"
+
+	"github.com/cosmos/cosmos-sdk/baseapp/config"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+type (
+	ExecMode uint8
+)
+
+const (
+	ExecModeCheck               = ExecMode(sdk.ExecModeCheck)               // Check a transaction
+	ExecModeReCheck             = ExecMode(sdk.ExecModeReCheck)             // Recheck a (pending) transaction after a commit
+	ExecModeSimulate            = ExecMode(sdk.ExecModeSimulate)            // Simulate a transaction
+	ExecModePrepareProposal     = ExecMode(sdk.ExecModePrepareProposal)     // Prepare a block proposal
+	ExecModeProcessProposal     = ExecMode(sdk.ExecModeProcessProposal)     // Process a block proposal
+	ExecModeVoteExtension       = ExecMode(sdk.ExecModeVoteExtension)       // Extend or verify a pre-commit vote
+	ExecModeVerifyVoteExtension = ExecMode(sdk.ExecModeVerifyVoteExtension) // Verify a vote extension
+	ExecModeFinalize            = ExecMode(sdk.ExecModeFinalize)            // Finalize a block proposal
+)
+
+type Manager struct {
+	// volatile states:
+	//
+	// - checkState is set on InitChain and reset on Commit
+	// - finalizeBlockState is set on InitChain and FinalizeBlock and set to nil
+	// on Commit.
+	//
+	// - checkState: Used for CheckTx, which is set based on the previous block's
+	// state. This state is never committed.
+	//
+	// - prepareProposalState: Used for PrepareProposal, which is set based on the
+	// previous block's state. This state is never committed. In case of multiple
+	// consensus rounds, the state is always reset to the previous block's state.
+	//
+	// - processProposalState: Used for ProcessProposal, which is set based on the
+	// the previous block's state. This state is never committed. In case of
+	// multiple rounds, the state is always reset to the previous block's state.
+	//
+	// - finalizeBlockState: Used for FinalizeBlock, which is set based on the
+	// previous block's state. This state is committed.
+	checkState           *State
+	prepareProposalState *State
+	processProposalState *State
+	finalizeBlockState   *State
+	stateMut             sync.RWMutex
+
+	gasConfig config.GasConfig
+}
+
+func NewManager(gasConfig config.GasConfig) *Manager {
+	return &Manager{
+		gasConfig: gasConfig,
+	}
+}
```

Comment:
the min gas prices value can be set on baseapp without the app being `sealed` which implies this value could be dynamic. i think we should use the value from the app when we need it, rather than storing it once and using the same value.

its entirely possible that devs or validators have a custom node impl that has dynamic minGasPrices.

Reply:
do you think it makes sense to just let folks set it once before being `sealed`? thats a semantic difference but probably makes more sense imo.  

The only time you change your min gas prices is when you edit your app toml or with a flag and restart your app

Reply:
yeah you're right nvm, when i looked yesterday i thought it was a setter, but its just an initialization option